R/utils.R

```diff
@@ -58,39 +58,59 @@ py_global_env <- function() {
   reticulate::py
 }
 
-# Create a duplicate of a Python environment
-#
-# @examples
-# reticulate::py_run_string("x = 3")
-# new_py_envir <- duplicate_py_env(py)
-# new_py_envir$items()
-#
-# @param module the `py` module accessed via `reticulate`
-#
-# @return a Python `Dict` or dictionary that is not converted to an R data type
-# @keywords internal
-duplicate_py_env <- function(module) {
+#' Create a duplicate of a Python environment
+#'
+#' @return a Python `Dict` or dictionary
+#'
+#' @noRd
+#' @keywords internal
+#'
+#' @examples
+#' \dontrun{
+#' reticulate::py_run_string("x = 3")
+#' new_py_envir <- duplicate_py_env()
+#' new_py_envir$items()
+#' }
+duplicate_py_env <- function() {
   rlang::check_installed("reticulate", "Python exercise support")
 
-  # extract all objects within this module
-  new_objs <- reticulate::py_get_attr(module, "__dict__")
-  # then create copy of the dictionary with all objects
-  copy <- reticulate::import("copy", convert = FALSE)
-  copy$copy(new_objs)
-}
-
-# This clears the Python environment `py`.
-#
-# It will keep important initial objects such as `py` (main module), 
-# `r` (reticulate interface to R), and the `builtins` module.
-#
-# @examples
-# reticulate::py_run_string("x = 3")
-# # this removes the `x`
-# clear_py_env()
-#
-# @return Nothing
-# @keywords internal
+  # grab Python utility functions
+  py_utilities <- py_maybe_load_utilities()
+
+  # extract all objects of `reticulate::py` (the main module)
+  py_env <- reticulate::py_get_attr(py_global_env(), "__dict__")
+
+  # deep copy Python environment
+  py_utilities$deep_copy(py_env)
+}
+
+py_maybe_load_utilities <- function() {
+  utilities <- reticulate::py[["__dict__"]][["__learnr__"]]
+  if (!is.null(utilities)) {
+    return(utilities)
+  }
+
+  reticulate::py_run_file(
+    system.file("internals", "learnr.py", package = "learnr"),
+    convert = FALSE
+  )
+}
+
+#' This clears the Python environment `py`.
+#'
+#' It will keep important initial objects such as `py` (main module),
+#' `r` (reticulate interface to R), and the `builtins` module.
+#'
+#' @noRd
+#' @keywords internal
+#' @return Nothing
+#'
+#' @examples
+#' \dontrun{
+#' reticulate::py_run_string("x = 3")
+#' # this removes the `x`
+#' clear_py_env()
+#' }
 clear_py_env <- function() {
```

Comment:
```suggestion
py_clear_env <- function() {
```

Reply:
Or maybe it should also include `global` in the name.

```suggestion
py_clear_global_env <- function() {
```